heapq.c

```diff
@@ -103,15 +104,48 @@ void heapq_push(struct heapq_t *hq, void *item)
 void heapq_free(struct heapq_t *hq)
 {
     free(hq->items);
-    free(hq);
 }
 
-struct heapq_t *heapq_alloc(nicc_hq_compare_func *cmp)
+void heapq_init(struct heapq_t *hq, compare_fn_t *cmp)
 {
-    struct heapq_t *hq = malloc(sizeof(struct heapq_t));
     hq->size = 0;
     hq->capacity = HEAPQ_STARTING_CAPACITY;
     hq->items = malloc(hq->capacity * sizeof(void *));
     hq->cmp = cmp;
-    return hq;
+}
+
+static void heap_sort_internal(u8 *left, u8 *right, size_t size, compare_fn_t cmp)
+{
+    struct heapq_t heap;
+    heapq_init(&heap, cmp);
+
+    u8 *p = left;
+    u8 elems[(right - left) / size][size];
+    u32 i = 0;
+    while (p <= right) {
+	memcpy(*(elems + i), p, size);
+	heapq_push(&heap, *(elems + i++));
+	p += size;
+    }
```

Comment:
Here it seems my brain may have seg-faulted...
```c
u8 elems[(right - left) / size][size];
u32 i = 0;
while (p <= right) {
  memcpy(*(elems + i), p, size);
  heapq_push(&heap, *(elems + i++));
  p += size;
}
```
<br>
Can be swapped with
```c
u8 tmp_elem[size];
while (p <= right) {
  memcpy(tmp_elem, p, size);
  heapq_push(&heap, tmp_elem));
  p += size;
}
```

What a baffling mistake to make...

Reply:
nvm this was good